src/lua/lua_tcp.c

```diff
@@ -1320,6 +1626,175 @@ lua_tcp_request (lua_State *L)
 	return 1;
 }
 
+/***
+ * @function rspamd_tcp.connect_sync({params})
+ * Creates new pseudo-synchronous connection to the specific address:port
+ *
+ * - `task`: rspamd task objects (implies `pool`, `session`, `ev_base` and `resolver` arguments)
+ * - `ev_base`: event base (if no task specified)
+ * - `resolver`: DNS resolver (no task)
+ * - `session`: events session (no task)
+ * - `config`: config (no task)
+ * - `host`: IP or name of the peer (required)
+ * - `port`: remote port to use
+ * - `timeout`: floating point value that specifies timeout for IO operations in **seconds**
+ * @return {boolean} true if request has been sent
+ */
+static gint
+lua_tcp_connect_sync (lua_State *L)
+{
+	LUA_TRACE_POINT;
+	GError *err = NULL;
+
+	gint64 port = -1;
+	gdouble timeout = default_tcp_timeout;
+	const gchar *host = NULL;
+	gint ret;
+	guint64 h;
+
+	struct rspamd_task *task = NULL;
+	struct rspamd_async_session *session = NULL;
+	struct rspamd_dns_resolver *resolver = NULL;
+	struct rspamd_config *cfg = NULL;
+	struct event_base *ev_base = NULL;
+
+	int arguments_validated = rspamd_lua_parse_table_arguments (L, 1, &err,
+			"task=U{task};session=U{session};resolver=U{resolver};ev_base=U{ev_base};"
+			"*host=S;*port=I;timeout=N;config=U{config}",
+			&task, &session, &resolver, &ev_base,
+			&host, &port, &timeout, &cfg);
+
+	if (!arguments_validated) {
+		if (err) {
+			ret = luaL_error (L, "invalid arguments: %s", err->message);
+			g_error_free (err);
+
+			return ret;
+		}
+
+		return luaL_error (L, "invalid arguments");
+	}
+
+	if (0 > port || port > 65535) {
+		return luaL_error (L, "invalid port given (correct values: 1..65535)");
+	}
+
+	if (task == NULL && (cfg == NULL || ev_base == NULL || session == NULL)) {
+		return luaL_error (L, "invalid arguments: either task or config+ev_base+session should be set");
+	}
+
+	if (timeout < 0.00001) {
```

Comment:
It looks weird. I'd suggest to set timeout to NaN, and switch `timeout=N` to `timeout=D` and then check for `isnan` here and set default timeout if it is NaN.